File: src/base58.c
// Header files
#include <alloca.h>
#include <string.h>
#include "base58.h"
#include "common.h"


// Definitions

// Number base
#define NUMBER_BASE 58

// Size percent increase
#define SIZE_PERCENT_INCREASE 138

// Size percent decrease
#define SIZE_PERCENT_DECREASE 733


// Constants

// Characters
static const char CHARACTERS[] = {'1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'J', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'};


// Function prototypes

// Base58 encode get length
static size_t base58EncodeGetLength(char *result, const uint8_t *data, const size_t length);

// Base58 decode get length
static size_t base58DecodeGetLength(uint8_t *result, const char *data, const size_t length);

// Get checksum
static void getChecksum(uint8_t *checksum, const uint8_t *data, const size_t length);


// Supporting function implementation

// Get base58 encoded length
size_t getBase58EncodedLength(const uint8_t *data, const size_t length) {

	// Return base58 encoded length
	return base58EncodeGetLength(NULL, data, length);
}

// Base58 encode
void base58Encode(char *result, const uint8_t *data, const size_t length) {

	// Get base58 encoding of the data
	base58EncodeGetLength(result, data, length);
}

// Get base58 encoded length with checksum
size_t getBase58EncodedLengthWithChecksum(const uint8_t *data, const size_t length) {

	// Return base58 encoded length with checksum
	return base58EncodeGetLength(NULL, data, length + BASE58_CHECKSUM_SIZE);
}

// Base58 encode with checksum
void base58EncodeWithChecksum(char *result, uint8_t *data, const size_t length) {

	// Append checksum to the data
	getChecksum(&data[length - BASE58_CHECKSUM_SIZE], data, length - BASE58_CHECKSUM_SIZE);

	// Get base58 encoding of the data
	base58EncodeGetLength(result, data, length);
}

// Get base58 decoded length
size_t getBase58DecodedLength(const char *data, const size_t length) {

	// Return base58 decoded length
	return base58DecodeGetLength(NULL, data, length);
}

// Base58 decode
void base58Decode(uint8_t *result, const char *data, const size_t length) {

	// Get base58 decoding of the data
	base58DecodeGetLength(result, data, length);
}

// Get base58 decoded length with checksum
size_t getBase58DecodedLengthWithChecksum(const char *data, const size_t length) {

	// Get base58 decoded length
	const size_t resultLength = base58DecodeGetLength(NULL, data, length);

	// Return base58 decoded length with checksum if valid
	return (resultLength >= BASE58_CHECKSUM_SIZE) ? resultLength : INVALID_BASE58_SIZE;
}

// Base58 decode with checksum
bool base58DecodeWithChecksum(uint8_t *result, const char *data, const size_t length) {

	// Get base58 decoding of the data
	const size_t resultLength = base58DecodeGetLength(result, data, length);

	// Get checksum of the result
	uint8_t checksum[BASE58_CHECKSUM_SIZE];
	getChecksum(checksum, result, resultLength - BASE58_CHECKSUM_SIZE);

	// Return if the checksum matches the expected result
	return !memcmp(&result[resultLength - BASE58_CHECKSUM_SIZE], checksum, sizeof(checksum));
}

// Base58 encode get length
size_t base58EncodeGetLength(char *result, const uint8_t *data, const size_t length) {

	// Go through all leading zeros in the data
	size_t numberOfLeadingZeros = 0;
	while(numberOfLeadingZeros < length && !data[numberOfLeadingZeros]) {

		// Increment number of leading zeros
		++numberOfLeadingZeros;
	}

	// Get buffer size
	const size_t bufferSize = (length - numberOfLeadingZeros) * SIZE_PERCENT_INCREASE / 100 + 1;

	// Create buffer
	uint8_t *buffer = alloca(bufferSize);
	explicit_bzero(buffer, bufferSize);

	// Go through all bytes in the byte array after the leading zeros
	size_t currentLength = 0;
	for(size_t i = numberOfLeadingZeros; i < length; ++i) {

		// Get byte
		uint8_t byte = data[i];

		// Go through all base 58 components of the byte
		size_t j = 0;
		for(size_t k = bufferSize - 1; byte || j < currentLength; --k) {

			// Include the current buffer value in the byte
			const uint16_t value = byte + (UINT8_MAX + 1) * buffer[k];

			// Set value in the buffer
			buffer[k] = value % NUMBER_BASE;

			// Update the byte
			byte = value / NUMBER_BASE;

			// Increment j
			++j;

			// Check if at the last component
			if(!k) {

				// Break
				break;
			}
		}

		// Update current length
		currentLength = j;
	}

	// Go through all leading zeros in the buffer
	size_t bufferIndex = bufferSize - currentLength;
	while(bufferIndex < bufferSize && !buffer[bufferIndex]) {

		// Increment buffer index
		++bufferIndex;
	}

	// Get result size
	const size_t resultSize = bufferSize - bufferIndex + numberOfLeadingZeros;

	// Check if getting result
	if(result) {

		// Set result to start with the number of leading zeros in base58
		memset(result, CHARACTERS[0], numberOfLeadingZeros);

		// Go through all bytes in the buffer after the leading zeros
		for(size_t i = 0; bufferIndex < bufferSize; ++i, ++bufferIndex) {

			// Append buffer's value in base58 to the result
			result[i + numberOfLeadingZeros] = CHARACTERS[buffer[bufferIndex]];
		}
	}

	// Return result size
	return resultSize;
}

// Base58 decode get length
size_t base58DecodeGetLength(uint8_t *result, const char *data, const size_t length) {

	// Go through all leading zeros in base58 in the string
	size_t numberOfLeadingZeros = 0;
	while(numberOfLeadingZeros < length && data[numberOfLeadingZeros] == CHARACTERS[0]) {

		// Increment number of leading zeros
		++numberOfLeadingZeros;
	}

	// Get buffer size
	const size_t bufferSize = (length - numberOfLeadingZeros) * SIZE_PERCENT_DECREASE / 1000 + 1;

	// Create buffer
	uint8_t *buffer = alloca(bufferSize);
	explicit_bzero(buffer, bufferSize);

	// Go through all characters in the string after the leading zeros in base58
	size_t currentLength = 0;
	for(size_t i = numberOfLeadingZeros; i < length; ++i) {

		// Get position of character in the characters
		const char *characterOffset = (char *)memchr(CHARACTERS, data[i], sizeof(CHARACTERS));

		// Check if character is invalid
		if(!characterOffset) {

			// Return invalid base58 size
			return INVALID_BASE58_SIZE;
		}

		// Get character as byte
		uint8_t byte = characterOffset - CHARACTERS;

		// Go through all base 58 components of the byte
		size_t j = 0;
		for(size_t k = bufferSize - 1; byte || j < currentLength; --k) {

			// Include the current buffer value in the byte
			const uint16_t value = byte + NUMBER_BASE * buffer[k];

			// Set value in the buffer
			buffer[k] = value % (UINT8_MAX + 1);

			// Update the byte
			byte = value / (UINT8_MAX + 1);

			// Increment j
			++j;

			// Check if at the last component
			if(!k) {

				// Break
				break;
			}
		}

		// Update current length
		currentLength = j;
	}

	// Go through all leading zeros in the buffer
	size_t bufferIndex = bufferSize - currentLength;
	while(bufferIndex < bufferSize && !buffer[bufferIndex]) {

		// Increment buffer index
		++bufferIndex;
	}

	// Get result size
	const size_t resultSize = bufferSize - bufferIndex + numberOfLeadingZeros;

	// Check if getting result
	if(result) {

		// Set result to start with the number of leading zeros
		explicit_bzero(result, numberOfLeadingZeros);

		// Go through all bytes in the buffer after the leading zeros
		for(size_t i = 0; bufferIndex < bufferSize; ++i, ++bufferIndex) {

			// Append buffer's value to the result
			result[i + numberOfLeadingZeros] = buffer[bufferIndex];
		}
	}

	// Return result size
	return resultSize;
}

// Get checksum
void getChecksum(uint8_t *checksum, const uint8_t *data, const size_t length) {

	// Get hash of data
	uint8_t hash[CX_SHA256_SIZE];
	cx_hash_sha256(data, length, hash, sizeof(hash));

	// Get hash of hash
	cx_hash_sha256(hash, sizeof(hash), hash, sizeof(hash));

	// Set checksum to hash
	memcpy(checksum, hash, BASE58_CHECKSUM_SIZE);
}

File: tests/functional_tests/base58.js
// Use strict
"use strict";


// Requires
const sha256 = require("./js-sha256-0.10.0.js");
const Common = require("./common.js");


// Classes

// Base58 class
class Base58 {

	// Public
		
		// Encode
		static encode(byteArray) {
		
			// Go through all leading zeros in the byte array
			var numberOfLeadingZeros = 0;
			while(numberOfLeadingZeros < byteArray["length"] && byteArray[numberOfLeadingZeros] === 0) {
			
				// Increment number of leading zeros
				++numberOfLeadingZeros;
			}
			
			// Get buffer size
			var bufferSize = Math.ceil((byteArray["length"] - numberOfLeadingZeros) * Base58.BYTES_PER_LENGTH_IN_NEW_BASE);
			
			// Create buffer
			var buffer = (new Uint8Array(bufferSize)).fill(0);
			
			// Go through all bytes in the byte array after the leading zeros
			var length = 0;
			for(var i = numberOfLeadingZeros; i < byteArray["length"]; ++i) {
			
				// Get byte
				var byte = byteArray[i];
				
				// Go through all base 58 components of the byte
				for(var j = 0, k = bufferSize - 1; (byte !== 0 || j < length) && k >= 0; ++j, --k) {
				
					// Include the current buffer value in the byte
					byte += (Common.BYTE_MAX_VALUE + 1) * buffer[k];
					
					// Set value in the buffer
					buffer[k] = byte % Base58.NUMBER_BASE;
					
					// Update the byte
					byte = Math.floor(byte / Base58.NUMBER_BASE);
				}
				
				// Update length
				length = j;
			}
			
			// Go through all leading zeros in the buffer
			var bufferIndex = bufferSize - length;
			while(bufferIndex < buffer["length"] && buffer[bufferIndex] === 0) {
			
				// Increment buffer index
				++bufferIndex;
			}
			
			// Set result to start with the number of leading zeros in base58
			var result = Base58.CHARACTERS[0].repeat(numberOfLeadingZeros);
			
			// Go through all bytes in the buffer after the leading zeros
			for(; bufferIndex < buffer["length"]; ++bufferIndex) {
			
				// Append buffer's value in base58 to the result
				result += Base58.CHARACTERS[buffer[bufferIndex]];
			}
			
			// Return result
			return result;
		}
		
		// Encode with checksum
		static encodeWithChecksum(byteArray) {
		
			// Get the checksum of the byte array
			var checksum = Base58.getChecksum(byteArray);
			
			// Return encoding the byte array with the checksum
			return Base58.encode(Common.mergeArrays([
					
				// Byte array
				byteArray,
				
				// Checksum
				checksum
			]));
		}
		
		// Decode
		static decode(base58String) {
		
			// Go through all leading zeros in base58 in the string
			var numberOfLeadingZeros = 0;
			while(numberOfLeadingZeros < base58String["length"] && base58String[numberOfLeadingZeros] === Base58.CHARACTERS[0]) {
			
				// Increment number of leading zeros
				++numberOfLeadingZeros;
			}
			
			// Get buffer size
			var bufferSize = Math.ceil((base58String["length"] - numberOfLeadingZeros) * Base58.BYTES_PER_LENGTH_IN_OLD_BASE);
			
			// Crete buffer
			var buffer = (new Uint8Array(bufferSize)).fill(0);
			
			// Go through all characters in the string after the leading zeros in base58
			var length = 0;
			for(var i = numberOfLeadingZeros; i < base58String["length"]; ++i) {
			
				// Get character as byte
				var byte = Base58.CHARACTERS.indexOf(base58String[i]);
				
				// Check if byte is invalid
				if(byte === Common.INDEX_NOT_FOUND) {
				
					// Throw error
					throw "Invalid base58 string.";
				}
				
				// Go through all base 58 components of the byte
				for(var j = 0, k = bufferSize - 1; (byte !== 0 || j < length) && k >= 0; ++j, --k) {
				
					// Include the current buffer value in the byte
					byte += Base58.NUMBER_BASE * buffer[k];
					
					// Set value in the buffer
					buffer[k] = byte % (Common.BYTE_MAX_VALUE + 1);
					
					// Update the byte
					byte = Math.floor(byte / (Common.BYTE_MAX_VALUE + 1));
				}
				
				// Update length
				length = j;
			}
			
			// Go through all leading zeros in the buffer
			var bufferIndex = bufferSize - length;
			while(bufferIndex < buffer["length"] && buffer[bufferIndex] === 0) {
			
				// Increment buffer index
				++bufferIndex;
			}
			
			// Set result to start with the number of leading zeros
			var result = (new Uint8Array(numberOfLeadingZeros + bufferSize - bufferIndex)).fill(0);
			
			// Go through all bytes in the buffer after the leading zeros
			for(var i = 0; bufferIndex < buffer["length"]; ++i, ++bufferIndex) {
			
				// Append buffer's value to the result
				result[i + numberOfLeadingZeros] += buffer[bufferIndex];
			}
			
			// Return result
			return result;
		}
		
		// Decode with checksum
		static decodeWithChecksum(base58String) {
		
			// Try
			try {
		
				// Decode the string
				var byteArray = Base58.decode(base58String);
			}
			
			// Catch errors
			catch(error) {
			
				// Throw error
				throw error;
			}
			
			// Check if the byte array doesn't include a checksum
			if(byteArray["length"] < Base58.CHECKSUM_LENGTH) {
			
				// Throw error
				throw "No checksum exists.";
			}
			
			// Otherwise
			else {
			
				// Get the checksum of the byte array without its checksum
				var checksum = Base58.getChecksum(byteArray.subarray(0, byteArray["length"] - Base58.CHECKSUM_LENGTH));
				
				// Get the provided checksum from the byte array
				var providedChecksum = byteArray.subarray(byteArray["length"] - Base58.CHECKSUM_LENGTH);
				
				// Check if checksums don't match
				if(Common.arraysAreEqual(checksum, providedChecksum) === false) {
				
					// Throw error
					throw "Invalid checksum.";
				}
				
				// Otherwise
				else {
				
					// Return byte array without the checksum
					return byteArray.subarray(0, byteArray["length"] - Base58.CHECKSUM_LENGTH);
				}
			}
		}
	
	// Private
	
		// Get checksum
		static getChecksum(byteArray) {
		
			// Get a hash of the hash of the byte array
			var hash = new Uint8Array(sha256.arrayBuffer(new Uint8Array(sha256.arrayBuffer(byteArray))));
			
			// Get the checksum from the hash
			var checksum = hash.subarray(0, Base58.CHECKSUM_LENGTH);
			
			// Return checksum
			return checksum;
		}
		
		// Characters
		static get CHARACTERS() {
		
			// Return characters
			return "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
		}
		
		// Number base
		static get NUMBER_BASE() {
		
			// Return number base
			return 58;
		}
		
		// Bytes per length in new base
		static get BYTES_PER_LENGTH_IN_NEW_BASE() {
		
			// Return bytes per length in new base
			return Math.log(Common.BYTE_MAX_VALUE + 1) / Math.log(Base58.NUMBER_BASE);
		}
		
		// Bytes per length in old base
		static get BYTES_PER_LENGTH_IN_OLD_BASE() {
		
			// Return bytes per length in old base
			return Math.log(Base58.NUMBER_BASE) / Math.log(Common.BYTE_MAX_VALUE + 1);
		}
		
		// Checksum length
		static get CHECKSUM_LENGTH() {
		
			// Return checksum length
			return 4;
		}
}


// Main function

// Set global object's base58
globalThis["Base58"] = Base58;

// Export base58
module["exports"] = Base58;
